### ISSUE TITLE
Stream size

### ISSUE THREAD

**Issue by fenichelar** [caolan/highland#501]
I have a stream that I am calling `slice()` on. I want the count before I call `slice()`. I am doing this:

    var total = 0;
    stream
    .map(function (object) {
      total++;
      return object;
    })
    .slice(2,5)

But `total` ends up being the size of the stream after the slice. How can I get the count correctly?

**Comment by vqvu** [comment]
The reason why this happens is because Highland streams are lazy, so they don't emit values (or execute callbacks) unless they have to. When you say `slice(2, 5)`, you're saying that you don't care about any element after the 5th one. Thus, the stream will simply not attempt to emit it.

In general, Highland streams don't support getting its size very well. For your specific case, there are two ways around this. 

If you can afford to materialize the entire stream, use `toArray` or `collect` to convert to an array and work with that.

If you can't, you'll need to implement your own `consume` callback that isn't lazy.

```javascript
var total = 0;
var i = 0;
stream.consume((err, x, push, next) => {
    if (err) {
        push(err);
        next();
    } else if (x === _.nil) {
        push(null, _.nil);
        // Do something with total.
    } else {
        if (start <= total && total < end) {
            push(null, x);
        }
        total++;
        next();
    }
});
```

**Comment by fenichelar** [comment]
Thanks! I ended up solving it by adding the `total++` to the `filter()` call.

On another note, this is what I am doing:

    highland(request({url: 'URL'}, function (error, response, body) {
      if (error || response.statusCode != 200) {
        deferred.reject(error);
      }
    }).pipe(converter))
    .map(function (object) {
      return JSON.parse(object.toString());
    })
    .filter(function(row) {
      var passed = searchRow(row, query.search);
      if (passed) total++;
      return passed;
    })
    .sortBy(sortByKey(query.sort, query.order))
    .slice(parseInt(query.offset || 0), parseInt(query.limit || data.length))
    .map(function (object) {
      object.id = hash(object);
      return object;
    })
    .toArray(function (data) {
      deferred.resolve({
        "total": total,
        "rows": data
      });
    });

Basically convert a csv into JSON and then filter it, sort it, and return it in pages. My hope was to be able to handle a response with millions of rows. Is Highland streams even a good idea? Just hoping to find out if I am going down a route that will never work well. Thanks! :)

**Comment by vqvu** [comment]
I think your approach may work, depending on how big the result set is. Typically Highland is useful if you want to do asynchronous operations or if you need the incremental operation of a stream (e.g., `array.map(...).filter(...)` is less efficient than `stream.map(...).filter(...)` since the former creates intermediate arrays).

In your case, Highland is useful up until you do `sortBy`. The problem with `sortBy` is that it will [materialize the entire dataset in memory](https://github.com/caolan/highland/blob/2.8.1/lib/index.js#L2977) before sorting it. As long as the query results are small enough, you'll be fine. But you could also end up with the entire millions of rows in memory if the search query doesn't filter out enough rows. There's nothing Highland can do to help you on this.

By the way, if you decide to go down this route, I suggest merging into a single `toArray` callback so you don't have to add a side-effect to your `filter` predicate. You want to avoid side effects in the `filter` or `map` callbacks, since it can be difficult to predict what will happen.

```javascript
stream.filter(row => searchRow(row, query.search))
    .toArray(data => {
        const sorted = data.sort(sortByKey(...))
        const result = sorted.slice(offset, limit)
            .map(object => {
               object.id = hash(object);
               return object;
            });
        deferred.resolve({
            total: sorted.length,
            rows: result
        })
    });
```

**Comment by fenichelar** [comment]
Thanks for the info! Unfortunately, most of the time the data will not have a filter key given. I'll probably try to come up with my own toArray method that builds the array while doing the sort and slice so that the entire array is never created, only the part I need. Keeping track of the index should make this possible. But it is currently working for about 1 million rows fairly quickly.

**Comment by vqvu** [comment]
Ah, ok. Sounds like you'll definitely need to create an index. I'm going to close this issue. Feel free to reopen it or create a new one if you have more questions.

**Comment by fenichelar** [comment]
```javascript
function orderByKey(key, reverse, a, b) {
  var x = (a[key] || '').toString().toLowerCase();
  var y = (b[key] || '').toString().toLowerCase();
  if (x > y) {
    return reverse ? -1 : 1;
  } else if (y > x) {
    return reverse ? 1 : -1;
  } else {
    return 0;
  }
}

function locationOf(element, array, key, reverse, start, end) {
  if (array.length === 0) {
    return -1;
  }

  start = start || 0;
  end = end || array.length;
  var pivot = parseInt(start + (end - start) / 2, 10);

  var c = orderByKey(key, reverse, element, array[pivot]);
  if (end - start <= 1) {
    return c == -1 ? pivot - 1 : pivot;
  }

  switch (c) {
    case -1: return locationOf(element, array, key, reverse, start, pivot);
    case 0: return pivot;
    case 1: return locationOf(element, array, key, reverse, pivot, end);
  }
}

var offset = parseInt(query.offset) || 0;
var limit = parseInt(query.limit) || Infinity;
var sortKey = query.sort;
var order = query.order === 'desc';

var total = 0;
var startIndex = 0;
var data = [];

highland(request({url: 'URL'}, function (error, response, body) {
  if (error || response.statusCode != 200) {
    deferred.reject(error);
  }
}).pipe(converter))
.map(function (object) {
  return JSON.parse(object.toString());
})
.filter(function(row) {
  return searchRow(row, query.search);
})
.each(function (row) {
  total++;
  if (data.length < limit) {
    array.splice(locationOf(row, data, sortKey, order) + 1, 0, element);
  } else {
    var index = locationOf(row, data, sortKey, order) + 1;
    if (startIndex < offset) {
      startIndex++;
      if (index > 0) {
        data.splice(index, 0, row);
        data.shift();
      }
    } else if (index < limit) {
      data.splice(index, 0, row);
      data.pop();
    }
  }
})
.done(function () {
  data.map(function (object) {
    object.id = hash(object);
    return object;
  });
  deferred.resolve({
    "total": total,
    "rows": data
  });
});
```

So I did it and it works well :). A good amount faster. Code above. Any idea what is blocking the event thread? Feel free to add a sorted/splice function to highlands using the code above!

**Comment by vqvu** [comment]
Nice! This seems sorted-splice seems a bit niche for the core library, but I'll add the "cookbook" label to make it easier to find this.

> Any idea what is blocking the event thread?

Maybe the calls to `splice` when you're constructing your `data` object or the calls to `locationOf` or both? You could potentially be inserting a lot of data in the middle of array. `data` probably isn't large, but you might still be doing it millions of times. And `locationOf` is called `nlog(limit)` times in total (where `n` is the size of your full dataset). For `n=1,000,000, limit=100`, that's something like 6.6 million calls on average.
